### PR TITLE
Python Challenge: array_product

### LINKED ISSUE #10: Challenge: array product
develop python function that gets an array as input, calculates the product of all elements in the give array except for the element at each index, and return the resulting array where it has the same length as the input array

### PR DESCRIPTION
**Pull request by Fahim-Ahmad**
- [x] PR's title includes the challenge's name and language
- [x] PR has short and clear description of the challenge
- [x] PR has appropriate labels and milestones for easy identification
- [x] PR it is assigned to the owner
- [x] reviewers are assigned
- [x] the PR contributes only one focused change
- [x] the branch is up to date with `main`/`master`
- [x] the code works when pulled and run locally
- [x] all conflicts are resolved (if any)
- [x] It is linked to an issue in the appropriate column of the project board
- [x] feedback is addressed (if any, and if it is appropriate feedback.)

## README Documentation

- [x] The solution is documented in `/src/README.md`
- [x] The markdown source is formatted
- [x] Spelling and grammar is correct in all text
- [x] The markdown looks correct when you preview the file
- [x] All links and images work
- [x] The README documents the solution's behavior, strategy(ies) and implementation(s)
- [x] There are use cases in code blocks to illustrate the function's behavior

## Python Files

- [ ] There is a module header
- [x] There is a module docstring
- [x] File names are in snake_case
- [x] Test files are named `test_<module_name>.py`

## Function Docstring

- [x] Behavior description
- [x] Parameter description
- [x] Return value description
- [x] Include any assumptions
- [x] Include 3 or more (passing!) doctests
- [x] Include 1-2 use cases (if necessary)

## Function Implementation

- [x] The solution is not is not copied from [djeada](https://github.com/djeada/Algorithms-And-Data-Structures)
- [x] The code is auto-formatted
- [x] The code has no (reasonable) linting mistakes
- [x] Variables are named with snake_case
- [x] The function has a clear and helpful name
- [x] The file's name matches the function name
- [x] The code follows the strategy as simply as possible
- [x] Variable names are clear and helpful
- [x] Comments explain the strategy (if necessary)
- [ ] There are type annotations
- [x] (challenge) The code includes defensive assertions

## Unit Test Suite

- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring
- Each unit test has
  - [x] A helpful name
  - [x] A clear docstring
  - [x] Only one assertion
- [x] All tests pass
- [x] There are tests for defensive assertions
- [x] There are tests for boundary cases
- [x] Test suite includes black-box and glass-box tests

### CODE REVIEW

File: src/brain_teasers/python/array_product/README.md [hunk 1]
```diff
@@ -0,0 +1,33 @@
+# Array Product Except Self
```
Comment (mohammedalasly) [inline src/brain_teasers/python/array_product/README.md]: You can change the document title to "Arry Product"  to match the Python file name

File: src/brain_teasers/python/array_product/README.md [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+This function takes an array as input and returns a new array of the same length where each element represents the product of all elements in the original array except the one at the corresponding index.
+
+```
```
Comment (mohammedalasly) [inline src/brain_teasers/python/array_product/README.md]: It will look better if you type python after the backtick, like  ```python

File: src/brain_teasers/python/array_product/README.md [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+product_of_array_except_self([])
+output: []
+
+product_of_array_except_self([1])
+output: [1]
+
+product_of_array_except_self([0])
+output: [0]
+
+```
```
Comment (mohammedalasly) [inline src/brain_teasers/python/array_product/README.md:33]: You can also improve your README to make it more well-documented if you add the command line that the user can use to run/test your code and a conclusion

**Review by mohammedalasly — CHANGES_REQUESTED** [review]

File: src/brain_teasers/python/array_product/src/array_product.py [hunk 4]
```diff
@@ -0,0 +1,34 @@
+def product_of_array_except_self(arr):
```
Comment (mohammedalasly) [inline src/brain_teasers/python/array_product/src/array_product.py]: Your code looks nice and it passed the test, It is only the module name that needs to be added on top, and matching the function name with the Python file name array_product

**Review by mohammedalasly — CHANGES_REQUESTED** [review]

File: src/brain_teasers/python/array_product/README.md (hunk 1 above)
Comment (Fahim-Ahmad) [reply src/brain_teasers/python/array_product/README.md]: Thanks. The title is now updated.

File: src/brain_teasers/python/array_product/README.md (hunk 2 above)
Comment (Fahim-Ahmad) [reply src/brain_teasers/python/array_product/README.md]: I liked it, thank you.
The word 'python' is now added after the backticks.

File: src/brain_teasers/python/array_product/src/array_product.py (hunk 4 above)
Comment (Fahim-Ahmad) [reply src/brain_teasers/python/array_product/src/array_product.py]: I should have thought about tit in advance. I changed the function name in to array_product and updated readme and unit-tests accordingly.

File: src/brain_teasers/python/array_product/README.md (hunk 3 above)
Comment (Fahim-Ahmad) [reply src/brain_teasers/python/array_product/README.md:33]: I agree. The readme file is now updated with the instruction for testing the code implementation.

**Review by mohammedalasly — APPROVED** [review]
Excellent, looks much better!!
